packages/SwingSet/src/kernel/collectionManager.js

```diff
@@ -0,0 +1,891 @@
+        !has(key),
+        X`key ${key} already registered in collection ${q(label)}`,
+      );
+      if (valueSchema) {
+        assert(
+          matches(value, valueSchema),
+          X`invalid value type for collection ${q(label)}`,
+        );
+      }
+      currentGenerationNumber += 1;
+      const serializedValue = serialize(value);
+      if (durable) {
+        serializedValue.slots.map(vref =>
+          assert(vrm.isDurable(vref), X`value is not durable`),
+        );
+      }
+      if (passStyleOf(key) === 'remotable') {
+        generateOrdinal(key);
+        const vref = getSlotForVal(key);
+        if (durable) {
+          assert(vrm.isDurable(vref), X`key is not durable`);
+        }
+        if (hasWeakKeys) {
+          vrm.addRecognizableValue(key, entryDeleter);
+        } else {
+          vrm.addReachableVref(vref);
+        }
+      }
+      serializedValue.slots.map(vrm.addReachableVref);
+      syscall.vatstoreSet(keyToDBKey(key), JSON.stringify(serializedValue));
+      ensureEntryCount();
+      entryCount += 1;
+    }
+
+    function set(key, value) {
+      assert(
+        matches(key, keySchema),
+        X`invalid key type for collection ${q(label)}`,
+      );
+      if (valueSchema) {
+        assert(
+          matches(value, valueSchema),
+          X`invalid value type for collection ${q(label)}`,
+        );
+      }
+      const dbKey = keyToDBKey(key);
+      const rawBefore = syscall.vatstoreGet(dbKey);
+      assert(rawBefore, X`key ${key} not found in collection ${q(label)}`);
+      const before = JSON.parse(rawBefore);
+      const after = serialize(harden(value));
+      if (durable) {
+        after.slots.map(vref =>
+          assert(vrm.isDurable(vref), X`value is not durable`),
+        );
+      }
+      vrm.updateReferenceCounts(before.slots, after.slots);
+      syscall.vatstoreSet(dbKey, JSON.stringify(after));
+    }
+
+    function deleteInternal(key) {
+      assert(
+        matches(key, keySchema),
+        X`invalid key type for collection ${q(label)}`,
+      );
+      const dbKey = keyToDBKey(key);
+      const rawValue = syscall.vatstoreGet(dbKey);
+      assert(rawValue, X`key ${key} not found in collection ${q(label)}`);
+      const value = JSON.parse(rawValue);
+      value.slots.map(vrm.removeReachableVref);
+      syscall.vatstoreDelete(dbKey);
+      let doMoreGC = false;
+      if (passStyleOf(key) === 'remotable') {
+        if (hasWeakKeys) {
+          vrm.removeRecognizableValue(key, entryDeleter);
+        } else {
+          doMoreGC = vrm.removeReachableVref(getSlotForVal(key));
+        }
+        deleteOrdinal(key);
+      }
+      ensureEntryCount();
+      entryCount -= 1;
+      return doMoreGC;
+    }
+
+    function del(key) {
+      currentGenerationNumber += 1;
+      deleteInternal(key);
+    }
+
+    function entriesInternal(
+      needValue,
+      keyPatt = M.any(),
+      valuePatt = M.any(),
+    ) {
+      assertKeyPattern(keyPatt);
```

Comment:
`assertKeyPattern` also checks that the pattern is frozen, right? I think it uses `passStyleOf`, so probably yes. Do we have any idea how expensive that check is? I lost track of code delegating to other code after about 6 hops.

Reply:
Yes.  No idea about cost.  We've got a big cost measurement thing looming in our future somewhere.